File: packages/sdks/everstack-node/src/resources/images.ts
/**
 * Images resource
 *
 * Provides image generation, editing, and variation functionality.
 */

import type { Client } from "@connectrpc/connect";
import type { GatewayService } from "@everstack/proto/everstack/gateway/v1/gateway_service_pb.js";
import { create } from "@bufbuild/protobuf";
import {
  ImageGenerationRequestSchema,
  type ImageGenerationResponse as ProtoImageGenResponse,
  ImageEditRequestSchema,
  type ImageEditResponse as ProtoImageEditResponse,
  ImageVariationRequestSchema,
  type ImageVariationResponse as ProtoImageVariationResponse,
} from "@everstack/proto/everstack/gateway/v1/images_pb.js";

import { fromConnectError } from "../errors.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface ImageGenerateParams {
  /** Text prompt describing the desired image */
  prompt: string;
  /** Model to use */
  model?: string;
  /** Number of images to generate (1-10) */
  n?: number;
  /** Image quality (standard, hd) */
  quality?: string;
  /** Response format (url, b64_json) */
  response_format?: string;
  /** Image size (256x256, 512x512, 1024x1024, 1792x1024, 1024x1792) */
  size?: string;
  /** Image style (vivid, natural) */
  style?: string;
  /** User identifier for abuse tracking */
  user?: string;
  /** Background setting */
  background?: string;
  /** Output format */
  output_format?: string;
  /** Moderation level */
  moderation?: string;
}

export interface ImageEditParams {
  /** The image to edit (PNG, max 4MB) */
  image: Uint8Array;
  /** Text prompt describing the desired edits */
  prompt: string;
  /** Optional mask image for inpainting */
  mask?: Uint8Array;
  /** Model to use */
  model?: string;
  /** Number of images to generate */
  n?: number;
  /** Image size */
  size?: string;
  /** Response format */
  response_format?: string;
  /** User identifier */
  user?: string;
  /** Image quality */
  quality?: string;
}

export interface ImageVariationParams {
  /** The image to create variations of */
  image: Uint8Array;
  /** Model to use */
  model?: string;
  /** Number of variations to generate */
  n?: number;
  /** Response format */
  response_format?: string;
  /** Image size */
  size?: string;
  /** User identifier */
  user?: string;
}

export interface ImageData {
  /** Base64-encoded image (if response_format is b64_json) */
  b64_json?: string;
  /** URL of the generated image (if response_format is url) */
  url?: string;
  /** The revised prompt used for generation */
  revised_prompt?: string;
}

export interface ImageUsage {
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
}

export interface ImageResponse {
  created: number;
  data: ImageData[];
  model: string;
  usage?: ImageUsage;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function transformImageData(
  proto: ProtoImageGenResponse | ProtoImageEditResponse | ProtoImageVariationResponse,
): ImageResponse {
  const usage = "usage" in proto && proto.usage
    ? {
        input_tokens: proto.usage.inputTokens,
        output_tokens: proto.usage.outputTokens,
        total_tokens: proto.usage.totalTokens,
      }
    : undefined;

  return {
    created: Number(proto.created),
    data: proto.data.map((d) => ({
      b64_json: d.b64Json || undefined,
      url: d.url || undefined,
      revised_prompt: d.revisedPrompt || undefined,
    })),
    model: proto.model,
    usage,
  };
}

// ---------------------------------------------------------------------------
// Resource
// ---------------------------------------------------------------------------

/**
 * Images resource for generation, editing, and variations
 */
export class Images {
  /** @internal */
  constructor(private readonly _client: Client<typeof GatewayService>) {}

  /**
   * Generate images from a text prompt
   *
   * @example
   * ```typescript
   * const response = await client.images.generate({
   *   prompt: 'A sunset over mountains',
   *   model: 'dall-e-3',
   *   size: '1024x1024',
   * });
   * console.log(response.data[0].url);
   * ```
   */
  async generate(params: ImageGenerateParams): Promise<ImageResponse> {
    try {
      const request = create(ImageGenerationRequestSchema, {
        prompt: params.prompt,
        model: params.model ?? "",
        n: params.n ?? 0,
        quality: params.quality ?? "",
        responseFormat: params.response_format ?? "",
        size: params.size ?? "",
        style: params.style ?? "",
        user: params.user ?? "",
        background: params.background ?? "",
        outputFormat: params.output_format ?? "",
        moderation: params.moderation ?? "",
      });

      const response = await this._client.imageGeneration(request);
      return transformImageData(response);
    } catch (error) {
      throw fromConnectError(error);
    }
  }

  /**
   * Edit an existing image with a text prompt
   *
   * @example
   * ```typescript
   * const response = await client.images.edit({
   *   image: imageBuffer,
   *   prompt: 'Add a rainbow in the sky',
   * });
   * ```
   */
  async edit(params: ImageEditParams): Promise<ImageResponse> {
    try {
      const request = create(ImageEditRequestSchema, {
        image: params.image,
        prompt: params.prompt,
        mask: params.mask ?? new Uint8Array(),
        model: params.model ?? "",
        n: params.n ?? 0,
        size: params.size ?? "",
        responseFormat: params.response_format ?? "",
        user: params.user ?? "",
        quality: params.quality ?? "",
      });

      const response = await this._client.imageEdit(request);
      return transformImageData(response);
    } catch (error) {
      throw fromConnectError(error);
    }
  }

  /**
   * Create variations of an existing image
   *
   * @example
   * ```typescript
   * const response = await client.images.createVariation({
   *   image: imageBuffer,
   *   n: 3,
   * });
   * ```
   */
  async createVariation(params: ImageVariationParams): Promise<ImageResponse> {
    try {
      const request = create(ImageVariationRequestSchema, {
        image: params.image,
        model: params.model ?? "",
        n: params.n ?? 0,
        responseFormat: params.response_format ?? "",
        size: params.size ?? "",
        user: params.user ?? "",
      });

      const response = await this._client.imageVariation(request);
      return transformImageData(response);
    } catch (error) {
      throw fromConnectError(error);
    }
  }
}

File: packages/sdks/everstack-python/src/everstack/resources/images.py
"""Images resource — generation, editing, variations."""

from __future__ import annotations

import base64
from typing import Any, Dict, Optional

from .._transport import Transport, AsyncTransport
from .._types import ImageResponse


class Images:
    def __init__(self, transport: Transport) -> None:
        self._transport = transport

    def generate(
        self,
        *,
        prompt: str,
        model: Optional[str] = None,
        n: Optional[int] = None,
        quality: Optional[str] = None,
        response_format: Optional[str] = None,
        size: Optional[str] = None,
        style: Optional[str] = None,
        user: Optional[str] = None,
        **kwargs: Any,
    ) -> ImageResponse:
        """Generate images from a text prompt."""
        body: Dict[str, Any] = {"prompt": prompt}
        if model:
            body["model"] = model
        if n is not None:
            body["n"] = n
        if quality:
            body["quality"] = quality
        if response_format:
            body["response_format"] = response_format
        if size:
            body["size"] = size
        if style:
            body["style"] = style
        if user:
            body["user"] = user
        body.update(kwargs)
        data = self._transport.request("POST", "/v1/images/generations", json_body=body)
        return ImageResponse.model_validate(data)

    def edit(
        self,
        *,
        image: bytes,
        prompt: str,
        mask: Optional[bytes] = None,
        model: Optional[str] = None,
        n: Optional[int] = None,
        size: Optional[str] = None,
        **kwargs: Any,
    ) -> ImageResponse:
        """Edit an existing image."""
        body: Dict[str, Any] = {
            "image": base64.b64encode(image).decode(),
            "prompt": prompt,
        }
        if mask:
            body["mask"] = base64.b64encode(mask).decode()
        if model:
            body["model"] = model
        if n is not None:
            body["n"] = n
        if size:
            body["size"] = size
        body.update(kwargs)
        data = self._transport.request("POST", "/v1/images/edits", json_body=body)
        return ImageResponse.model_validate(data)

    def create_variation(
        self,
        *,
        image: bytes,
        model: Optional[str] = None,
        n: Optional[int] = None,
        size: Optional[str] = None,
        **kwargs: Any,
    ) -> ImageResponse:
        """Create variations of an image."""
        body: Dict[str, Any] = {"image": base64.b64encode(image).decode()}
        if model:
            body["model"] = model
        if n is not None:
            body["n"] = n
        if size:
            body["size"] = size
        body.update(kwargs)
        data = self._transport.request("POST", "/v1/images/variations", json_body=body)
        return ImageResponse.model_validate(data)


class AsyncImages:
    def __init__(self, transport: AsyncTransport) -> None:
        self._transport = transport

    async def generate(self, *, prompt: str, **kwargs: Any) -> ImageResponse:
        body: Dict[str, Any] = {"prompt": prompt, **kwargs}
        data = await self._transport.request("POST", "/v1/images/generations", json_body=body)
        return ImageResponse.model_validate(data)

    async def edit(self, *, image: bytes, prompt: str, **kwargs: Any) -> ImageResponse:
        body: Dict[str, Any] = {
            "image": base64.b64encode(image).decode(),
            "prompt": prompt,
            **kwargs,
        }
        data = await self._transport.request("POST", "/v1/images/edits", json_body=body)
        return ImageResponse.model_validate(data)

    async def create_variation(self, *, image: bytes, **kwargs: Any) -> ImageResponse:
        body: Dict[str, Any] = {"image": base64.b64encode(image).decode(), **kwargs}
        data = await self._transport.request("POST", "/v1/images/variations", json_body=body)
        return ImageResponse.model_validate(data)
